### ISSUE TITLE
Use numpydoc pre-commit

### ISSUE THREAD

**Issue by nstarman** [astropy/astropy#15386]
### What is the problem this feature will solve?

We use numpydoc. But the actual docstrings aren’t checked for correctness.

### Describe the desired outcome

Their new pre-commit hook means docstrings are checked as part of the CI. 

### Additional context

https://github.com/numpy/numpydoc/issues/450

**Comment by eerovaher** [comment]
Is this supposed to replace [`docformatter`](https://github.com/PyCQA/docformatter)?

**Comment by nstarman** [comment]
They shouldn't be mutually incompatible, but yes that would be the hope. `numpydoc` has many more specific checks.

**Comment by namurphy** [comment]
I tried out numpydoc on PlasmaPy recently and ran into a problem (https://github.com/numpy/numpydoc/issues/519) where the crashed with an exception, rather than noting that the problem was with a specific rule. I'm waiting until this issue gets fixed before adding numpydoc to pre-commit.

This particular error happens when a line in a `See Also` section starts with a `~`, ``` `~ ```, or `|`.  I ran `validate-docstring` on Astropy's source code just now and ran into the same issue.

**Comment by mhvk** [comment]
Ah, that explains what happened at https://github.com/numpy/numpy/pull/25416/files#r1434272271 - I added a comment to your `numpydoc` issue as well.

**Comment by pllim** [comment]
> But the actual docstrings aren’t checked for correctness.

What kind of correctness? We already have RTD set to turn warning into errors. Is that not enough?

**Comment by astrofrog** [comment]
This will do things like check that all arguments are documented, etc

**Comment by tactipus** [comment]
so how can we solve this problem? What is the solution?

**Comment by nstarman** [comment]
> so how can we solve this problem? What is the solution?

Numpydoc has a pre-commit we can use. Numpydoc uses a config file specifying what error codes to check. We'd want to start by turning off any that we are currently failing then sequentially turn them on and fix the issues.

**Comment by astrofrog** [comment]
Yes we use it in reproject and it has found countless real issues so I am very much in favour of doing this. It is going to be a lot of work to fix issues!

**Comment by pllim** [comment]
But would this be obsolete with the typing stuff taking over in the future?

**Comment by nstarman** [comment]
In short, this would not be made obsolete by typing because numpydoc checks docstrings, which typing does not interact with. Longer answer is in 3-4 years kinda because Python is adding the `typing.Doc` annotation and it will be possible to build descriptive docstrings from type annotations, but even then there will be room for Numpydoc.

**Comment by namurphy** [comment]
I hadn't heard about `typing.Doc` but it sounds cool!  Here's the draft of [PEP 727](https://peps.python.org/pep-0727/) which is proposing `typing.Doc` with a target version of Python 3.13.

**Comment by nstarman** [comment]
Yeah, I've played around with it and it is interesting. On the downside, basically all function signatures become 1 parameter per line because of the attached string description. On the upsides: 1) docstrings are guaranteed to be in sync; 2) IDEs can surface descriptive information, not just type hints, about a Parameter on hover-over, 3) it's easier to switch docstring formats, like `numpy` to `google` (though I'm not sure why one would want to switch, but it is easier).

**Comment by mhvk** [comment]
Had a look too and really like the idea of doc'd arguments (and class variables, allowing us to remove the sphinx-only docstrings under them). Indeed, I'd go as far as this being something that would actually convince me to use typing myself... (and maybe to have `typing_extensions` as a required dependency...)

I think for one-line argument docs, the length issue is OK, but for multi-line I'm less sure. More generally, the really descriptive part of an argument, of how it influences the function/method, may still belong in the docstring -- this may need some thought (sadly).

Given that using `Doc` may well lead to a bit of a rethink of docstrings, I think the case for having numpydoc pre-commit is weakened; we don't want to go through such mini-edits multiple times.

**Comment by pllim** [comment]
Follow-up issue is https://github.com/astropy/astropy/issues/20353
